go/processor/floria/processor.go

```diff
@@ -70,31 +73,74 @@ func (p *processor) Run(
 		// Create new contract
 	} else {
 		// Call existing contract
-		result, err = call(p.interpreter, transaction, context)
+		result, err = call(p.interpreter, transaction, context, gas)
 		if err != nil {
-			return errorReceipt, nil
+			return tosca.Receipt{GasUsed: transaction.GasLimit}, nil
 		}
 	}
 
+	gasUsed := gasUsed(transaction, result.GasLeft)
+
 	return tosca.Receipt{
 		Success:         result.Success,
-		GasUsed:         transaction.GasLimit,
+		GasUsed:         gasUsed,
 		ContractAddress: nil,
 		Output:          result.Output,
 		Logs:            nil,
 	}, nil
 }
 
+func gasUsed(transaction tosca.Transaction, gasLeft tosca.Gas) tosca.Gas {
+	if transaction.Sender != (tosca.Address{}) {
```

Comment:
This special case handling for the zero-sender is surprising and at least note-worthy. Why would this be done?

If we confirm this, we should have an integration making sure that the surplus-gas charge is indeed not charged when a transaction was send by Address zero.

Reply:
The comment in the reference implementation says "For whatever reason". I added a comment stating that this is indented behavior, as well as an integration test.

Reply:
I remember writing the "whatever reason" comment :)
It was and still is a surprise.